### ISSUE TITLE
[ASPagerNode] Touches Began Not Called

### ISSUE THREAD

**Issue by ardaerzin** [facebook/AsyncDisplayKit#1419]
When using an ASPagerNode, touch related methods (began, moved etc) are not being called, but everything works as expected when I remove the pager. is this the desired behavior?

**Comment by hannahmbanana** [comment]
@ardaerzin - what object are you looking for touch related methods on? Have you checked that user interaction is enabled? It is default NO on all ASControlNodes, which includes text, image and button nodes, unless you call addTarget: or set userInteractionEnabled to YES.

Overall, could you add more detail to this issue? It would help to have your node / view hierarchy. Also, the version of the framework you are using. 

If you look at the examples/VerticalWithinHorizontal example app, it has an ASPagerNode with touch enabled working.

**Comment by ardaerzin** [comment]
@hannahmbanana sorry for the late response, timezone diff.

Basically I'm using an ASPagerNode to display 3 viewcontrollers that I have, pretty standard stuff. Nodes for the viewcontrollers are created using the `pagerNode(pagerNode: ASPagerNode!, nodeBlockAtIndex index: Int)` and everything works fine. 

The problem I'm having is not related to the elements in those views because I use them like you mentioned. One of the views have an editabletextnode in them, and I wanted to use `touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?)` on its parent node to dismiss the keyboard if the textnode has focus, but the method is never called. If I ditch the pager and present the VC alone, `touchesBegan` is called normally.

**Comment by appleguy** [comment]
@ardaerzin - I'm really sorry you are having this issue and that it took me a week to see this! If you could submit a sample project showing this behavior, I would be happy to debug it. Feel free to email me at AsyncDisplayKit@gmail.com if the sample contains sensitive material. Alternatively, there are several example projects that would be easy to modify for this scenario.

If you could try setting scrollEnabled = NO on the pager node's view and retest, that would be helpful debug information for me.

**Comment by ardaerzin** [comment]
@appleguy sorry I was overwhelmed by some deadlines. I created a very [basic project](https://github.com/ardaerzin/ASDKTest). I believe this should be enough to spot the issue. With a normal UIScrollView touchesBegan is called when scrollEnabled is set to false, but that does not work when I use an ASPagerNode.

**Comment by hannahmbanana** [comment]
@ardaerzin - Thanks for including a sample project, I apologize for the delay in response. We have been busy releasing AsyncDisplayKit 1.9.8, so make sure to check it out! @appleguy and I looked at your code and we found that it was referencing some older code styles of AsyncDisplayKit, which implies you are not on the latest version. You may need to run pod update to get the latest version of AsyncDisplayKit, as newer versions of CocoaPods require this. 

As for your question: To detect a tap, you should either add a tap gesture recognizer to the `ASCellNode` in `didLoad:` (_not_ in `init:`) or add an `ASControlNode` that is as large as the area you need to tap with an `addTarget:action:` call. 

Because your sample project was mostly empty, your current hierarchy seemed to have elements that weren't used or weren't necessary for the functionality of the sample project. The ChildControllerNode class was not instantiated anywhere (instead ASCellNode was created directly in the ViewController data source method) . Also you were manually creating an ASPagerNode layout, which did not seem to be necessary or fully configured, so because the only apparent customization to the ASPagerNode that the subcalss ViewControllerNode implemented was an orange background color, we simply created the ASPagerNode directly (rather than subclassing) and set the background color to orange. 

One more important note is that your nodeBlockAtIndex function was not thread safe. This code is called asynchronously and so accessing shared state like subcontrollers from within the block is not okay! It would be easy to change this to nodeAtIndex: to not have to worry about thread safety. 

Here's the project cleaned up a bit:
[ASDKTest-master 2.zip](https://github.com/facebook/AsyncDisplayKit/files/324772/ASDKTest-master.2.zip)

**Comment by appleguy** [comment]
@ardaerzin Hannah is correct - after her analysis I looked over the entire project with her, made some changes, etc.  Overall it would be great if you want to look at the changes we made, clean up any of the remaining unused code, and let us know if you still have questions.  It's not very clear to me why you would want touchesBegan on the pager node, because just like in a collection view in UIKit, the content inside the cells would normally block the touches from reaching that point.  Also touchesBegan should never be used directly to detect taps; a tap is not considered recognized until the touch lifts, and without too much distance in the finger moving.  The tap gesture recognizer or addTarget/action is the way to go.
